### DISCUSSION TITLE
Renovate performs an uncontrolled exit with 0 status and stops processing further repos while resolving terraform dependency

### DISCUSSION THREAD

**Discussion by dlouzan** [renovatebot/renovate#10812] [Request Help]
**How are you running Renovate?**

- [ ] WhiteSource Renovate hosted app on github.com
- [x] Self hosted

If using the hosted app, please skip to the next section.
Otherwise, if self-hosted, please complete the following:

Please select which platform you are using:

- [ ] Azure DevOps (dev.azure.com)
- [ ] Azure DevOps Server
- [ ] Bitbucket Cloud (bitbucket.org)
- [ ] Bitbucket Server
- [ ] Gitea
- [ ] github.com
- [ ] GitHub Enterprise Server
- [ ] gitlab.com
- [x] GitLab self-hosted 14.0.4

Renovate version: 25.52.5 (gitlab template v1.46.8)

**Describe the bug**

This seems https://github.com/renovatebot/renovate/issues/6299 again caused by using a proxy on a corporate network, but on a different data source, in this case with a terraform dependency that is not available.

Running on the corporate environment where the proxy is set:

```sh
 INFO: Dependency extraction complete (repository=code-ops/terraform-provider-itam)
       "baseBranch": "main",
       "stats": {
         "managers": {
           "gitlabci": {"fileCount": 1, "depCount": 2},
           "gomod": {"fileCount": 1, "depCount": 1},
           "npm": {"fileCount": 1, "depCount": 3},
           "terraform": {"fileCount": 1, "depCount": 3}
         },
         "total": {"fileCount": 4, "depCount": 9}
       }
DEBUG: Dependency node has unsupported value fermium (repository=code-ops/terraform-provider-itam)
DEBUG: terraform-provider.getDependencies() (repository=code-ops/terraform-provider-itam)
       "lookupName": "code-ops/itam"
DEBUG: getLabels(https://index.docker.io, library/golang, latest) (repository=code-ops/terraform-provider-itam)

# Exits at this point with 0, stopping further processing
```

On localhost where no proxy is set, the resolution of the dependency fails and processing continues:

```sh
 INFO: Dependency extraction complete (repository=code-ops/terraform-provider-itam)
       "baseBranch": "main",
       "stats": {
         "managers": {
           "gitlabci": {"fileCount": 1, "depCount": 2},
           "gomod": {"fileCount": 1, "depCount": 1},
           "npm": {"fileCount": 1, "depCount": 3},
           "terraform": {"fileCount": 1, "depCount": 3}
         },
         "total": {"fileCount": 4, "depCount": 9}
       }
DEBUG: Dependency node has unsupported value fermium (repository=code-ops/terraform-provider-itam)
DEBUG: terraform-provider.getDependencies() (repository=code-ops/terraform-provider-itam)
       "lookupName": "code-ops/itam"
DEBUG: getLabels(https://index.docker.io, library/golang, latest) (repository=code-ops/terraform-provider-itam)
DEBUG: getManifestResponse(https://index.docker.io, library/golang, latest) (repository=code-ops/terraform-provider-itam)
DEBUG: getManifestResponse(https://index.docker.io, library/golang, sha256:264ebcddca05426de1c8a04cfb3a27fff72468c1d32b2d0f367d614e11449362) (repository=code-ops/terraform-provider-itam)
DEBUG: Datasource connection error (repository=code-ops/terraform-provider-itam)
       "datasource": "terraform-provider",
       "lookupName": "code-ops/itam",
       "url": undefined,
       "errCode": "ENOTFOUND"
DEBUG: Failed to look up dependency itam (repository=code-ops/terraform-provider-itam, packageFile=example/main.tf, dependency=itam)

# Processing continues
```

**Relevant debug logs**

<details><summary>Click me to see logs</summary>

```

```

</details>

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [x] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

...

/cc @max-wittig @bufferoverflow @ercanucan @fgreinacher

**Comment by dlouzan** [comment]
Actually just confirmed on the corporate environment, unsetting the proxy variables fails the dependency resolution (as expected), but renovate does not die and continues processing.

**Reply by viceice** [reply]
What do you expect? does it work with a simple public reproduction repo?

**Reply by rarkins** [reply]
Could be a global-agent error? Strange to have an unlogged and immediate exit

**Reply by viceice** [reply]
I think this is an issue i've seen before at our corporate environment without proxy, but using http2. Node simply died silently on http2 connections.

Do you use the `enableHttp2` host rule setting?

https://docs.renovatebot.com/configuration-options/#enablehttp2

**Reply by dlouzan** [reply]
@viceice Thanks for the hint, but unfortunately we do not use that setting :-(

**Reply by viceice** [reply]
Then maybe your corporate proxy is somehow misbehaving, so  that the `global-agent` or `got` fail with a hidden error.

I think there isn't much we can do. We would need a reproduction for it:
- create a proxy like yours
- create a simple nodejs script with `global-agent` and `got` and try to do the failing request.
- create an issue upstream (`global-agent`, `got`, or `proxy`)

**Reply by rarkins** [reply]
Is there any type of uncaught exception we can add, or is this type of error too "hard" to catch?

**Reply by viceice** [reply]
I've tried with known node handlers, but never had success to catch it. 😕

**Reply by rarkins** [reply]
I guess this is reason to avoid "monkey patching" node or libraries which do. I wonder if we use any apart from global-agent and got which do in production?

**Reply by dlouzan** [reply]
Trace level for renovate does not log anything. I'll try fiddling with the node logger settings, maybe something pops up at the trace level, I remember doing something similar in the past.

**Reply by dlouzan** [reply]
> Then maybe your corporate proxy is somehow misbehaving

Oh, I'm sure our proxy misbehaves. It always misbehaves.

**Reply by dlouzan** [reply]
I might have a good lead here. Unfortunately neither the `DEBUG: *` nor the `NODE_DEBUG: *` variables helped much, all output seems legit, but after reading this upstream issue https://github.com/nodejs/node/issues/22088 I modified my docker container renovate launch script adding a couple of handlers:

```node
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("/usr/src/app/node_modules/source-map-support/register");
const logger_1 = require("/usr/src/app/node_modules/renovate/dist/logger");
if (process.argv[1] !== '/usr/local/bin/renovate') {
    logger_1.logger.warn("Deprecated! Please use global 'renovate' command.");
}

process.on('uncaughtException', (exception) => {
  console.log(exception);
});

process.on('unhandledRejection', (reason, p) => {
  console.log("Unhandled Rejection at: Promise ", p, " reason: ", reason);
});

process.on('beforeExit', (code) => {
  console.log("I'm actually out of events in the loop: ", code);
});

require("/usr/src/app/node_modules/renovate/dist/renovate");
//# sourceMappingURL=renovate.js.map
```

And the output of this is:

```sh
 INFO: Dependency extraction complete (repository=code-ops/terraform-provider-itam)
       "baseBranch": "main",
       "stats": {
         "managers": {
           "gitlabci": {"fileCount": 1, "depCount": 2},
           "gomod": {"fileCount": 1, "depCount": 1},
           "npm": {"fileCount": 1, "depCount": 3},
           "terraform": {"fileCount": 1, "depCount": 3}
         },
         "total": {"fileCount": 4, "depCount": 9}
       }
DEBUG: Dependency node has unsupported value fermium (repository=code-ops/terraform-provider-itam)
DEBUG: terraform-provider.getDependencies() (repository=code-ops/terraform-provider-itam)
       "lookupName": "code-ops/itam"
DEBUG: getLabels(https://index.docker.io, library/golang, latest) (repository=code-ops/terraform-provider-itam)
DEBUG: getManifestResponse(https://index.docker.io, library/golang, latest) (repository=code-ops/terraform-provider-itam)
DEBUG: getManifestResponse(https://index.docker.io, library/golang, sha256:264ebcddca05426de1c8a04cfb3a27fff72468c1d32b2d0f367d614e11449362) (repository=code-ops/terraform-provider-itam)
I'm actually out of events in the loop:  0
```

So this doesn't look like a crash, it's a controlled exit. According to the linked upstream issue, could this be a wrong handling of a promise when the connection fails using the proxy code path?

According to https://nodejs.org/api/process.html#process_event_beforeexit:

> The 'beforeExit' event is emitted when Node.js empties its event loop and has no additional work to schedule. Normally, the Node.js process will exit when there is no work scheduled, but a listener registered on the 'beforeExit' event can make asynchronous calls, and thereby cause the Node.js process to continue.

**Reply by viceice** [reply]
So it looks like the issue i had with http2. So the server and nodejs can't properly handle something, which causes node to silently exit, because it thinks the event loop is empty.

So i think it's not a renovate issue, but a upstream issue 😕

**Reply by fgreinacher** [reply]
Have you already [enabled global-agent logging](https://github.com/gajus/global-agent#enable-logging)? Maybe that provides some additional information.

**Reply by dlouzan** [reply]
@fgreinacher As usual very helpful 😁 That might just have been the final touch, this is a *very* suspicious thing I see after enabling those logs:

```json
{
  "context": {
    "package":"global-agent",
    "namespace":"Agent",
    "logLevel":10,
    "destination":"https://[gitlab-domain]undefinedcode-ops/itam",
    "proxy":"[corporate-proxy]",
    "requestId":2
  },
  "message":"proxying request",
  "sequence":11,
  "time":1626277676317,
  "version":"1.0.0"
}
```

Notice the literal `undefined` in the destination url; this makes the processing fail and go through the proxy when it shouldn't since it doesn't match the `no_proxy`, and also cause an internal failure due to the invalid url.

**Reply by rarkins** [reply]
Bug in our terraform providers datasource or an incorrect registryUrls config?

**Reply by dlouzan** [reply]
I don't see anything wrong in the provider definition, apart from it not being available (expected, currently not deployed):

```hcl
terraform {
  required_providers {
    itam = {
      version = "0.4.0"
      source  = "[gitlab-domain]/code-ops/itam"
    }
  }
}
```

**Reply by rarkins** [reply]
What does it look like in the "packageFiles with updates" log?

**Reply by dlouzan** [reply]
As far as I can see, it never reaches that part of the log, the process exits before calculating the updates. The end of the log is what I pasted above.

**Comment by fgreinacher** [comment]
Seems that the [latest `global-agent` release](https://github.com/gajus/global-agent/releases/tag/v2.2.0) brought support for multiple entries in `NO_PROXY` (which applies in our environment).

My MR https://github.com/renovatebot/renovate/pull/6348 fixed the NuGet manager to not query local/relative URLs. Maybe `global-agent` has a general problem when querying such URLs. And maybe one of the (now properly evaluated) entries in our `NO_PROXY` configuration is interpreted in a similar way. 

@dlouzan Could you reset `NO_PROXY` and see whether it's still happening then?

**Reply by dlouzan** [reply]
I tried setting `NO_PROXY` to a single entry for our main domain, I observed the same behaviour.

For fully unsetting `NO_PROXY`, unfortunately I cannot set the proxy variables while at the same time removing the exceptions, in that case the query to our GitLab would then go through the proxy and this actually does not work.

**Comment by dlouzan** [comment]
Is there any reason this was moved to a discussion? Not that I have an issue, but seems a bit weird ❓

**Reply by rarkins** [reply]
I think let's move it back to being treated as a bug/issue

**Reply by dlouzan** [reply]
@rarkins Did I misunderstand your message or you were planning to convert this into a bug again for tracking? :bow:

**Reply by rarkins** [reply]
I just converted it in #11208
